lib/systems/src/controllers/DrivebrainController.cpp

```diff
@@ -0,0 +1,74 @@
+    auto last_torque_lim_timestamp = db_input.torque_limits.last_recv_millis;
+
+    // 2 if the time between the current VCR eval_millis time and the last millis time that we recvd a drivebrain msg is too high
+    bool speed_setpoint_msg_too_latent = (::abs((int)(static_cast<int64_t>(eval_millis) - static_cast<int64_t>(last_speed_setpoint_timestamp))) > (int)_params.allowed_latency);
+    bool torque_limit_message_too_latent = (::abs((int)(static_cast<int64_t>(eval_millis) - static_cast<int64_t>(last_torque_lim_timestamp))) > (int)_params.allowed_latency);
+
+    // std::cout << "last_speedsp ts " << last_speed_setpoint_timestamp <<std::endl;
+    // std::cout << "ts " << last_torque_lim_timestamp <<std::endl;
+    // 3 if the relative latency is too high (time between the message members) -> (allowed latency / 2)
+    bool latency_diff_too_high = (::abs((int)(static_cast<int64_t>(last_speed_setpoint_timestamp) - static_cast<int64_t>(last_torque_lim_timestamp))) > ((int)_params.allowed_latency / 2));
+    
+    constexpr int64_t debug_timestamp_period_ms = 5000;
+
+    if((static_cast<int64_t>(eval_millis) - static_cast<int64_t>(_last_worst_latency_timestamp)) > debug_timestamp_period_ms)
+    {    
+        _last_worst_latency_timestamp = eval_millis;
+        _sub_message_latencies = {-1, -1};
+    }
+
+    if( (eval_millis - last_speed_setpoint_timestamp) > _sub_message_latencies.worst_speed_setpoint_latency_so_far)
+    {
+        _sub_message_latencies.worst_speed_setpoint_latency_so_far = (eval_millis - last_speed_setpoint_timestamp);   
+    } else if((eval_millis - last_torque_lim_timestamp) > _sub_message_latencies.worst_torque_lim_latency_so_far)
+    {
+        _sub_message_latencies.worst_torque_lim_latency_so_far = (eval_millis - last_torque_lim_timestamp);   
+    }
+
+    bool timing_failure = (speed_setpoint_msg_too_latent || torque_limit_message_too_latent || not_all_messages_recvd || latency_diff_too_high);
+    // if(timing_failure)
+    // {
+    //     std::cout <<"timing failures: "<<std::endl;
+    //     std::cout << speed_setpoint_msg_too_latent<< std::endl;
+    //     std::cout << torque_limit_message_too_latent<< std::endl;
+    //     std::cout << not_all_messages_recvd<< std::endl;
+    //     std::cout << latency_diff_too_high<< std::endl;
+    // }
+    // only if this is being evaluated while not the active control mode do we clear fault
+    bool is_active_controller = state.system_data.tc_mux_status.active_controller_mode == _params.assigned_controller_mode;
+
+    if ((!is_active_controller) && (!timing_failure))
+    {
+        // timing failure should be false here
+        _timing_failure = false;
+    }
+
+    DrivetrainCommand_s output;
+    if (!timing_failure && (!_timing_failure))
+    {   
+        output = db_input.get_command();
+    }
+    else
+    {
+        
+        _timing_failure = true;
+        output = _emergency_control.evaluate(state, eval_millis);
+    }
+    return output;
+}
```

Comment:
Is there any way to exit the emergency control mode? I don't see a way for _timing_failure to be set to false again.

Reply:
just a question about intended implementation-- I'm fine either way, just want to know